### PR TITLE
cache blackhole and passthrough clusters

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
These two clusters only depend on meshconfig.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[x] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

File: pilot/pkg/networking/core/v1alpha3/configgen.go [hunk 1]
```diff
@@ -24,16 +24,24 @@ type ConfigGeneratorImpl struct {
...
 }
 
 // Called when mesh config is changed.
-func (configgen *ConfigGeneratorImpl) MeshConfigChanged(_ *meshconfig.MeshConfig) {
+func (configgen *ConfigGeneratorImpl) MeshConfigChanged(mesh *meshconfig.MeshConfig) {
 	accessLogBuilder.reset()
+
+	// Reset blackhole and passthrough clusters
+	cb := NewClusterBuilder(nil, &model.PushContext{Mesh: mesh})
+	configgen.CDSCache.SetBlackHoleCluster(cb.buildBlackHoleCluster())
```
Comment (howardjohn) [inline pilot/pkg/networking/core/v1alpha3/configgen.go:45]: I think this can race with BuildClusters. Not a -race detected race, but we can accidentally revert the changes.

**Review by howardjohn — COMMENTED** [review]
Slightly worried about the complexity of this for only caching 2 clusters. Its not that bad now but will get even more complex once we fix the race? Maybe we can do it with https://github.com/istio/istio/pull/33338?

Or at least a subset of that PR, changing the BuildClusters to return discovery.Resource so we can skip marshaling in the configgen?

File: pilot/pkg/networking/core/v1alpha3/configgen.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/configgen.go:45]: There is a lock in CDSCache.

**Comment by hzxuzhonghu** [comment]
I will test  #33338 to have a look at what's the performance improve

File: pilot/pkg/networking/core/v1alpha3/configgen.go (hunk 1 above)
Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/configgen.go:45]: Yes its not a data race but you compute the result THEN acquire the lock. Without computing the resource inside the lock you may write out of order. See the "token" stuff in XdsCache it was done to solve similar issue

Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/configgen.go:45]: the concern is: compute v1, compute V2, write V2, write v1

**Comment by ramaraochavali** [comment]
I think better to do benchmark to see if the complexity is worth. I have tried in the past, but it did not give significant benefits so left it.  Is n't this automatically handled by cluster caching?

**Comment by hzxuzhonghu** [comment]
Yes,  if we have all cds cache, then it is not needed.
